Updated graph endpoint docs

ikas opened the pull request:
Related to [GTC-740](https://gfw.atlassian.net/browse/GTC-740).

tiagojsag reviewed (changes requested):
I feel like we need a clearer explanation of what a "concept" is - perhaps even in the "concepts" section, and not so much here. We mention it multiple times in these docs, but I think we never really explain what it is.

ethanroday commented on source/includes/api/_graph.md:
```diff
@@ -1,583 +1,865 @@
+
+We assume that, at this point, you are already familiar with the concepts of [dataset](#dataset), [layer](#layer), [widget](#widget) or [metadata](#metadata) (ideally you are familiar with the 4). If that's not the case, we strongly encourage you to learn about those concepts first, and we also recommend you take some time to explore and experiment using actual [dataset](#dataset6), [layer](#laye ...
+
+In the context of the RW API Graph service, resources are represented as graph nodes. They can be associated (using graph edges) with concepts to create relationships of relevance between the two entities.
```
Excellent summary of what this service does!

ethanroday commented on source/includes/api/_graph.md:
```diff
@@ -1,583 +1,865 @@
+This endpoint returns the list of concepts available in the graph. If successful, the response will have status 200 OK, containing a list of elements in the `data` index, each containing the information about one concept. Check out the [Graph concept reference](#graph-concept-reference) for details on each of the fields of the returned response.
 
-This endpoint returns the set of concepts that are inferred from the set passed as a parameter
+Please keep in mind this endpoint **does not return a paginated list** - instead, it returns all the concepts available in the graph, no matter how many they are. You should avoid using this specific endpoint since it might harm the performance of your application, should the graph increase significantly in size.
```
What should I use instead? Also, if we're basically asking users not to use this endpoint, we should maybe make that more prominent.

ikas replied:
Clarified this with a bold paragraph at the beginning of the section :)

ethanroday commented on source/includes/api/_graph.md:
```diff
@@ -1,583 +1,865 @@
+---------- | ------------- | ---------------------------------
+401        | Unauthorized  | You must provide your API user's token in the response headers.
+
+## Increment dataset view count
```
Doesn't this mean I could just spam this endpoint? I feel like this should be authenticated

ikas replied:
Unfortunately, the way things are right now, yes.

I agree this should be restricted to authenticated users (or even to other services, maybe), but that's out of the scope of the current work. Also, changing this to an authenticated endpoint is a breaking change, which will/might cause existing requests to start failing... Not saying we shouldn't do it **(on the contrary, I completely agree it should be done)**, but we should be careful about it.

tiagojsag replied:
I think there's a valid point to this endpoint being open to anonymous users - maybe this is meant to be used in a google analytics style, where we aim to have a global count of user interest, wether or not said users have accounts. We'd have to see where it's used (probably RW) and see if we can infer the motivation from there.

ethanroday commented on source/includes/api/_graph.md:
```diff
@@ -1,583 +1,865 @@
+------------------------ | -------------- | ---------------------------------------------------------------------------- 
+`id`                     | String         | The concept unique identifier.
+`label`                  | String         | A readable version of the concept identifier.
+`synonyms`               | String | Array | The list of synonyms for this concept (or an empty string, if no synonyms exist).
```
Table syntax error here - the pipe in `String | Array` puts array in the third column and cuts off the description

ethanroday reviewed (changes requested):
Overall, looks good! I do think it could use a little bit more elaboration at the beginning about how the graph actually gets constructed and what the end-to-end process is when I, say, want to associate a dataset with some concepts. Specific things that tripped me up:
 - Where do the concepts come from? Is there some kind of custom conceptual hierarchy? Or do users contribute concepts?
 - The whole idea of the concept hierarchy is never fully explained - what does it mean that there are "ancestor concepts"?
 - Are _none_ of these endpoints paginated? If so, we should point that out.
 - I would make a note about the automatic creation/deletion of resources in the graph at the beginning so the user doesn't think they'll have to manage this whole thing themselves.

ikas commented:
Hey @ethanroday, thanks a lot for the feedback :bow: 

I have introduced a new "bird's-eye view" section at the beginning, aiming at clarifying some of the points you mentioned - hope it helps get a clearer view. Please have a look and let me know your thoughts :) 

Also, a note about concept hierarchy: there are in fact some references to concept hierarchy across the docs - some avoidable and can just be removed in favor of alternatives (i.e. the "Query similar datasets including descendants" endpoint), some not so much (the "default_parent_id" field included in many responses). **At the moment, there's no hierarchy of concepts whatsoever.** I believe these references might have been added in the past with the intention of building concept hierarchy, but those features were not completed. I added a clarifying note about this in the initial bird's-eye view section, let me know if you think it is enough, or if we should aim at remove any reference about concept hierarchy from the docs.

ethanroday commented on source/includes/api/_graph.md:
````diff
@@ -43,9 +53,9 @@ curl -X GET https://api.resourcewatch.org/v1/graph/query/list-concepts
 ```
 
-This endpoint returns the list of concepts available in the graph. If successful, the response will have status 200 OK, containing a list of elements in the `data` index, each containing the information about one concept. Check out the [Graph concept reference](#graph-concept-reference) for details on each of the fields of the returned response.
+**This endpoint does not return a paginated list - you will receive a full list of all the concepts available in the graph. Additionally, the returned response can only be filtered by application, not searched - which means that it's very unlikely that you will receive a reduced payload by using this endpoint. As such, uou should avoid using this specific endpoint since it might harm the performa ...
````
```suggestion
**This endpoint does not return a paginated list - you will receive a full list of all the concepts available in the graph. Additionally, the returned response can only be filtered by application, not searched - which means that it's very unlikely that you will receive a reduced payload by using this endpoint. As such, you should avoid using this specific endpoint since it might harm the performance of your application, should the graph increase significantly in size. You can find alternative endpoints in the sections below that allow you to navigate the graph by similarity of concepts, which is the recommended approach to navigating the graph.**
```

ethanroday reviewed (approved):
Birds-eye view section is very helpful!

As far as the increment view count goes, perhaps we can consider that as a use case for API keys and rate limiting in the future.
